webapp/dojo_news/src/App.spec.js

```diff
@@ -0,0 +1,221 @@
+import {shallowMount} from '@vue/test-utils';
+import App from "./App";
+import 'regenerator-runtime';
+import Vue from "vue";
+
+describe("News list", () => {
+    it('does not render any items on an initially empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+    });
+
+    it('renders items on an initially filled list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                    {id: 1, title: "Linux", votes: 0},
+                    {id: 2, title: "Windows", votes: 0}
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+
+        expect(wrapper.findAll(".newslistitem").length).toBe(3);
+    });
+    it('renders items correctly dependent on mutations of the list of NewsItems', async () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                ]
+            }
+        });
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(1);
+
+        wrapper.vm.removeNewsListItem(0);
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+
+        wrapper.vm.createNewsListItem("Test");
+        wrapper.vm.createNewsListItem("Test2");
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(2);
+    });
+});
```

Comment:
There are 2 things that are being tested here, adding items and removing items. I think it would be better to separate those into two test cases so that it will be more clear where a test fails.

Reply:
Yes, and also DRY your test case descriptions a bit:
```
  News list
    ✓ does not render any items on an initially empty list of NewsItems (29 ms)
    ✓ renders items on an initially filled list of NewsItems (8 ms)
    ✓ renders items correctly dependent on mutations of the list of NewsItems (12 ms)
  List Empty Message
    ✓ displays a message on an initially empty list of NewsItems (3 ms)
    ✓ displays no message on an initially non-empty list of NewsItems (4 ms)
    ✓ adds and removes the message on list mutations (6 ms)
  Reverse Order Button
    ✓ list order is reversed upon Reverse-Button click from initial state _descending_ (8 ms)
    ✓ list order is reversed upon Reverse-Button click from initial state _ascending_ (6 ms)
    ✓ displays the Reverse-Order-Button on an initially non empty list of NewsItems (5 ms)
    ✓ does not display the Reverse-Order-Button on an initially empty list of NewsItems (4 ms)
    ✓ Reverse-Button is hidden upon empty list (6 ms)
    ✓ Reverse-Button shows up again upon non-empty list (6 ms)
```

What about something like:
```
  News list
    given no items
      renders empty state
```
```

Reply:
While I agree with you on writing two seperate tests for two functionalities, I think the structure of the tests is more of a personal preference thing. We structured our tests by feature, so we add the state to each of the test case titles. You would structure it by state, but then we would have to add the feature title to it. Or is there any other reason for our version to be bad practice?

Reply:
Just make sure your test case descriptions are documenting the behaviour of your code well. It's always a little tricky because it depends on the perception of humans.

I draw lot of my knowledge from sites like https://www.betterspecs.org/

![Screenshot - 2020-11-15T140650 199](https://user-images.githubusercontent.com/2110676/99185703-d8e74e00-274b-11eb-9b06-c31c898d5aa3.png)

But yes, I like your idea of structuring your tests by feature.
```
  News list
    emty state
      is displayed when list items are empty
```

Reply:
I definitely see your point, it's a dream to be able to read code and/or test results like that. I renamed all the test cases for a better reading experience of the `npm run test:unit` output.

But I still think that adding another description layer would be too verbose in this case. Because for example, you would have the description "given no items" for every feature, just for one test case. And if I were to do it like that, I also had to add the descriptions "given items", "upon empty news list" and "upon non-empty news list". Everywhere, just for one test case, which in my eyes ist way too verbose.